File: mobile/lib/features/camera/rep_counter.dart
import 'camera_state.dart';

/// Per-exercise configuration for rep counting
class ExerciseConfig {
  final String primaryAngleKey;
  final double downThreshold;
  final double upThreshold;
  final bool useSymmetry; // Average left and right sides
  final bool isTimerBased; // For plank
  final double? minHoldSeconds; // For plank

  const ExerciseConfig({
    required this.primaryAngleKey,
    required this.downThreshold,
    required this.upThreshold,
    this.useSymmetry = false,
    this.isTimerBased = false,
    this.minHoldSeconds,
  });
}

/// Rep counter with per-exercise phase detection
class RepCounter {
  String _currentExerciseId = '';
  RepPhase _phase = RepPhase.lockout;
  int _reps = 0;
  DateTime? _phaseStartTime;
  DateTime? _plankStartTime;

  static const double _minRepDuration = 0.5; // 500ms minimum per rep

  /// Exercise-specific configurations
  static const Map<String, ExerciseConfig> _configs = {
    'squat': ExerciseConfig(
      primaryAngleKey: 'knee_flexion',
      downThreshold: 100,
      upThreshold: 160,
      useSymmetry: true,
    ),
    'deadlift': ExerciseConfig(
      primaryAngleKey: 'hip_flexion',
      downThreshold: 70,
      upThreshold: 160,
      useSymmetry: true,
    ),
    'bench-press': ExerciseConfig(
      primaryAngleKey: 'elbow_extension',
      downThreshold: 80,
      upThreshold: 155,
      useSymmetry: true,
    ),
    'push-up': ExerciseConfig(
      primaryAngleKey: 'elbow_extension',
      downThreshold: 90,
      upThreshold: 155,
      useSymmetry: true,
    ),
    'pull-up': ExerciseConfig(
      primaryAngleKey: 'elbow_extension',
      downThreshold: 75, // Inverted: down is extended
      upThreshold: 155, // Up is flexed
      useSymmetry: true,
    ),
    'overhead-press': ExerciseConfig(
      primaryAngleKey: 'elbow_extension',
      downThreshold: 90,
      upThreshold: 165,
      useSymmetry: true,
    ),
    'lunges': ExerciseConfig(
      primaryAngleKey: 'knee_flexion',
      downThreshold: 95,
      upThreshold: 160,
      useSymmetry: true,
    ),
    'bicep-curl': ExerciseConfig(
      primaryAngleKey: 'elbow_extension',
      downThreshold: 55,
      upThreshold: 155,
      useSymmetry: true,
    ),
    'tricep-extension': ExerciseConfig(
      primaryAngleKey: 'elbow_extension',
      downThreshold: 80,
      upThreshold: 155,
      useSymmetry: true,
    ),
    'rows': ExerciseConfig(
      primaryAngleKey: 'elbow_extension',
      downThreshold: 80,
      upThreshold: 150,
      useSymmetry: true,
    ),
    'lat-pulldown': ExerciseConfig(
      primaryAngleKey: 'elbow_extension',
      downThreshold: 80,
      upThreshold: 155,
      useSymmetry: true,
    ),
    'leg-press': ExerciseConfig(
      primaryAngleKey: 'knee_flexion',
      downThreshold: 90,
      upThreshold: 160,
      useSymmetry: true,
    ),
    'plank': ExerciseConfig(
      primaryAngleKey: 'spine_angle',
      downThreshold: 160,
      upThreshold: 180,
      isTimerBased: true,
      minHoldSeconds: 30.0,
    ),
    'shoulder-press': ExerciseConfig(
      primaryAngleKey: 'elbow_extension',
      downThreshold: 90,
      upThreshold: 165,
      useSymmetry: true,
    ),
  };

  int get reps => _reps;
  RepPhase get phase => _phase;

  /// Initialize for a specific exercise
  void initialize(String exerciseId) {
    _currentExerciseId = exerciseId;
    reset();
  }

  /// Reset counter state
  void reset() {
    _reps = 0;
    _phase = RepPhase.lockout;
    _phaseStartTime = DateTime.now();
    _plankStartTime = null;
  }

  /// Update with new angle data
  /// Returns true when a rep is completed
  bool update(Map<String, double> angles) {
    final config = _configs[_currentExerciseId];
    if (config == null) return false;

    // Special handling for plank (timer-based)
    if (config.isTimerBased) {
      return _updatePlank(angles, config);
    }

    // Get primary angle (average left/right if needed)
    final angle = _getPrimaryAngle(angles, config);
    if (angle == null) return false;

    return _updatePhase(angle, config);
  }

  /// Get primary angle for the exercise
  double? _getPrimaryAngle(Map<String, double> angles, ExerciseConfig config) {
    if (!config.useSymmetry) {
      return angles[config.primaryAngleKey];
    }

    // Average left and right sides
    final leftKey = 'left_${config.primaryAngleKey}';
    final rightKey = 'right_${config.primaryAngleKey}';
    
    final leftAngle = angles[leftKey];
    final rightAngle = angles[rightKey];

    if (leftAngle == null && rightAngle == null) return null;
    if (leftAngle == null) return rightAngle;
    if (rightAngle == null) return leftAngle;

    return (leftAngle + rightAngle) / 2;
  }

  /// Update phase state machine
  bool _updatePhase(double angle, ExerciseConfig config) {
    final now = DateTime.now();
    bool completed = false;

    // Check minimum rep duration
    if (_phaseStartTime != null) {
      final phaseDuration = now.difference(_phaseStartTime!).inMilliseconds / 1000;
      if (phaseDuration < _minRepDuration && _phase != RepPhase.lockout) {
        return false; // Too fast, ignore
      }
    }

    // Handle pull-up (inverted logic)
    final bool isPullUp = _currentExerciseId == 'pull-up';
    final effectiveAngle = isPullUp ? 180 - angle : angle;
    final effectiveDown = isPullUp ? 180 - config.upThreshold : config.downThreshold;
    final effectiveUp = isPullUp ? 180 - config.downThreshold : config.upThreshold;

    switch (_phase) {
      case RepPhase.lockout:
        if (effectiveAngle < effectiveDown) {
          _phase = RepPhase.down;
          _phaseStartTime = now;
        }
        break;

      case RepPhase.down:
        // Stay in down phase briefly
        if (now.difference(_phaseStartTime!).inMilliseconds > 100) {
          _phase = RepPhase.pause;
        }
        break;

      case RepPhase.pause:
        if (effectiveAngle > effectiveDown) {
          _phase = RepPhase.up;
          _phaseStartTime = now;
        }
        break;

      case RepPhase.up:
        if (effectiveAngle > effectiveUp) {
          _phase = RepPhase.lockout;
          _phaseStartTime = now;
          _reps++;
          completed = true;
        }
        break;
    }

    return completed;
  }

  /// Update plank (timer-based)
  bool _updatePlank(Map<String, double> angles, ExerciseConfig config) {
    final spineAngle = angles['spine_angle'];
    if (spineAngle == null) return false;

    final now = DateTime.now();
    final isGoodForm = spineAngle >= config.downThreshold && 
                       spineAngle <= config.upThreshold;

    if (isGoodForm) {
      _plankStartTime ??= now;
      
      final holdDuration = now.difference(_plankStartTime!).inSeconds;
      if (holdDuration >= (config.minHoldSeconds ?? 30)) {
        // Complete one "rep" every 30 seconds of good plank
        _reps++;
        _plankStartTime = now; // Reset timer for next rep
        return true;
      }
    } else {
      _plankStartTime = null; // Reset if form breaks
    }

    return false;
  }

  /// Get current exercise configuration
  ExerciseConfig? getConfig() {
    return _configs[_currentExerciseId];
  }
}

File: backend/inference_service/app/ml/rep_counter.py
"""Rep Counting — phase-based, not frame-classification-based.

"Rep counting should not rely solely on frame classification. Instead use
movement phases: Down -> Pause -> Up -> Lockout. One complete cycle equals
one repetition." (technical specification)
"""
from enum import Enum, auto


class RepPhase(Enum):
    LOCKOUT = auto()
    DOWN = auto()
    PAUSE = auto()
    UP = auto()


# Exercise-specific phase thresholds, keyed by the primary tracked angle.
# Extend this table when adding new exercises — no inference logic changes
# required, per "Extensions should be modular" in the spec.
EXERCISE_THRESHOLDS = {
    "squat": {"angle": "left_knee_flexion", "down": 100, "up": 160},
    "deadlift": {"angle": "left_hip_flexion", "down": 90, "up": 160},
    "bench_press": {"angle": "left_elbow_extension", "down": 80, "up": 160},
    "push_up": {"angle": "left_elbow_extension", "down": 90, "up": 160},
    "bicep_curl": {"angle": "left_elbow_extension", "down": 160, "up": 50},
    "overhead_press": {"angle": "left_elbow_extension", "down": 90, "up": 165},
}


class RepCounter:
    def __init__(self, exercise: str) -> None:
        config = EXERCISE_THRESHOLDS.get(exercise, EXERCISE_THRESHOLDS["squat"])
        self.angle_key = config["angle"]
        self.down_threshold = config["down"]
        self.up_threshold = config["up"]
        self.phase = RepPhase.LOCKOUT
        self.reps = 0
        self.rep_angle_log: list[list[float]] = []
        self._current_rep_angles: list[float] = []

    def update(self, angles: dict[str, float]) -> bool:
        """Feed one frame's joint angles. Returns True exactly when a rep completes."""
        angle = angles.get(self.angle_key)
        if angle is None:
            return False

        self._current_rep_angles.append(angle)
        completed = False
        inverted = self.down_threshold > self.up_threshold  # e.g. bicep curl

        def passed_down() -> bool:
            return angle < self.down_threshold if not inverted else angle > self.down_threshold

        def passed_up() -> bool:
            return angle > self.up_threshold if not inverted else angle < self.up_threshold

        if self.phase == RepPhase.LOCKOUT and passed_down():
            self.phase = RepPhase.DOWN
        elif self.phase == RepPhase.DOWN:
            self.phase = RepPhase.PAUSE
        elif self.phase == RepPhase.PAUSE and passed_down():
            self.phase = RepPhase.UP
        elif self.phase == RepPhase.UP and passed_up():
            self.phase = RepPhase.LOCKOUT
            self.reps += 1
            completed = True
            self.rep_angle_log.append(self._current_rep_angles)
            self._current_rep_angles = []

        return completed
